### ISSUE TITLE
Build outputs from 'make' in release & debug build modes can get mixed-up, resulting in unstable binaries.

### ISSUE THREAD

**Issue by gapisback** [vmware/splinterdb#374]
### Describe the bug

In the dev workflow, one might tend to do `make` (for a fully optimized release build) and then inadvertently do a `make debug` to generate a debug build. With support for sanitized builds, the chances that we might build the library and tests with mixed set of objects is quite possible.

Devs have encountered very unstable test runs due to these types of accidents. We usually end up with strange seg-faults, leading to waste dev-hours trying to debug issues arising due to bad builds.

### Reproduction steps

```bash
1. Try this sequence: make
2. Followed by : make debug, without an intervening 'make clean'
3.
...

It's not exactly clear what causes these bad builds, but certainly a missing ` make clean ` seems to the source of the problem.
```

### Expected behavior

Solution proposed: We need some tight checking mechanism in `Makefile` to verify the state of an existing build, and if any build-parameters (CFLAGS, build-mode, env-vars that affect sanitized builds) change, `make` should flag it down, and refuse to build mixed-mode builds. Instruct the developer to do a `make clean`, and restart from a full clean-build.

### Additional context

A solution for this has been prototyped, and shown to be promising. This issue is being logged to externalize this developer productivity issue, for better tracking & visibility with an upcoming PR.
